File: 2017/06/Day6.py
from typing import List


def redistribute_memory_pt1(memory: List[int]):
    blocks = max(memory)
    index = memory.index(blocks)
    memory[index] = 0
    while blocks != 0:
        index += 1
        if index >= len(memory):
            index -= len(memory)
        memory[index] += 1
        blocks -= 1
    return memory


def do_reallocate_pt1(memory: List[int]):
    seen_states = []
    while memory not in seen_states:
        seen_states.append(memory)
        memory = redistribute_memory_pt1(memory[:])
    return len(seen_states)


def redistribute_memory_pt2(memory: List[int]):
    blocks = max(memory)
    index = memory.index(blocks)
    memory[index] = 0
    while blocks != 0:
        index += 1
        if index >= len(memory):
            index -= len(memory)
        memory[index] += 1
        blocks -= 1
    return memory


def do_reallocate_pt2(memory: List[int]):
    seen_states = []
    while memory not in seen_states:
        seen_states.append(memory)
        memory = redistribute_memory_pt2(memory[:])
    return len(seen_states) - seen_states.index(memory)


input_data = list(map(int, open("input.txt").read().split("\t")))

print(do_reallocate_pt1(input_data[:]))
print(do_reallocate_pt2(input_data[:]))

File: 2017/06/Day6.fs
module AdventOfCode.Solutions.Y2017.D06

open System.IO
open AdventOfCode.Solutions.Utils

let inputData =
    File.ReadAllText "input.txt" |> String.split "\t" |> Array.map int

[<TailCall>]
let rec redistribute
    (index: int)
    (blocks: int)
    (memory: int array)
    : int array =
    if blocks = 0 then
        memory
    else
        redistribute
            ((index + 1) % memory.Length)
            (blocks - 1)
            (memory |> Array.updateAt index (memory[index] + 1))

[<TailCall>]
let rec performReallocation
    (memory: int array)
    (seen: int array array)
    : int array * int array array =
    if seen |> Array.contains memory then
        memory, seen
    else
        let maxBlocks = memory |> Array.max
        let maxIndex = memory |> Array.findIndex ((=) maxBlocks)

        let updatedMemory =
            memory
            |> Array.updateAt maxIndex 0
            |> redistribute ((maxIndex + 1) % memory.Length) maxBlocks

        performReallocation updatedMemory (seen |> Array.add memory)

let solve () =
    let memory, seen = performReallocation inputData [||]
    let seenCount = seen |> Array.length

    seenCount |> printfn "%d"

    let firstSeenIndex = seen |> Array.findIndex ((=) memory)

    seenCount - firstSeenIndex |> printfn "%d"
